### PR TITLE
Feature/nested-outline

### PR DESCRIPTION
**Pull request by kv9898**
Fixes https://github.com/posit-dev/positron/issues/3822.

With the following code:
```r
# first level ####
a <- 1 # missing symbol
###########################
## second level ####
### third level####
some_function <- function(x){
  b <- 2
  # inside first level ####
  ## inside second level####
}
```

The outline is now displayed as:

image

### CODE REVIEW

**Comment by kv9898** [comment]
What's still missing is the foldable feature

**Review by lionel- — COMMENTED** [review]
Good first stab! This will be such a nice improvement.

Unfortunately we can't use unsafe code in this way, and especially not if safety is not established from a thorough analysis.

It seems it should be possible to move the comment section handling _inside_ the loop over the children of a node. And swap out the current parent there depending on the level. Then you shouldn't need to pass this stack around and deal with the ensuing reference lifetime issues.

WDYT?

File: crates/ark/src/lsp/symbols.rs [hunk 1]
```diff
@@ -218,22 +250,21 @@ fn index_assignment(
...
+    // otherwise, just index as a generic variable/object
     let name = contents.node_slice(&lhs)?.to_string();
 
     let start = convert_point_to_position(contents, lhs.start_position());
     let end = convert_point_to_position(contents, lhs.end_position());
 
     let symbol = DocumentSymbol {
         name,
-        kind: SymbolKind::OBJECT,
+        kind: SymbolKind::VARIABLE, // Use VARIABLE kind to represent an assignment
```
[the remark this answers is not in the document]
Comment (kv9898) [reply crates/ark/src/lsp/symbols.rs:267]: You are right! I used AI code assistant to rewrite this part and it changed the symbol (the blue box looked nice to me though). I will revert back to `OBJECT`.

File: crates/ark/src/lsp/symbols.rs [hunk 2]
```diff
@@ -218,22 +250,21 @@ fn index_assignment(
...
     let rhs = node.child_by_field_name("rhs").into_result()?;
 
-    // check for identifier on lhs, function on rhs
-    let function = lhs.is_identifier_or_string() && rhs.is_function_definition();
-
-    if function {
-        return index_assignment_with_function(node, contents, parent, symbols);
+    // check if lhs is an identifier and rhs is a function definition
+    if lhs.is_identifier_or_string() && rhs.is_function_definition() {
+        // If the RHS is a function definition, we call `index_assignment_with_function`
```
[the remark this answers is not in the document]
Comment (kv9898) [reply crates/ark/src/lsp/symbols.rs:255]: Sure, I will delete it.

File: crates/ark/src/lsp/symbols.rs [hunk 3]
```diff
@@ -141,58 +151,79 @@ fn index_node(
...
+                if *current_level >= level {
+                    section_stack.pop();
+                } else {
+                    break;
+                }
+            }
+
+            // Push the new symbol to the current parent
+            if let Some((_, current_parent_ptr)) = section_stack.last() {
+                let current_parent = unsafe { &mut **current_parent_ptr };
```
[the remark this answers is not in the document]
Comment (kv9898) [reply crates/ark/src/lsp/symbols.rs:186]: You are right! Just didn't work out how to do it in a safe way.

**Comment by kv9898** [comment]
> Good first stab! This will be such a nice improvement.
> 
> Unfortunately we can't use unsafe code in this way, and especially not if safety is not established from a thorough analysis.
> 
> It seems it should be possible to move the comment section handling _inside_ the loop over the children of a node. And swap out the current parent there depending on the level. Then you shouldn't need to pass this stack around and deal with the ensuing reference lifetime issues.
> 
> WDYT?

That's very clever, I feel like it will be much neater this way! I guess we can write the method for comment section handling first and plug it in the loop?

This is my first thought for the loop:
 1. if the node is detected as a comment, we call the new method for comment handling and do not call `index_node`,
 2. if the node is of other types, we simply recursively call `index_node` as we previously do.

Do you think this would work?

File: crates/ark/src/lsp/symbols.rs [hunk 4]
```diff
@@ -136,41 +136,69 @@ fn parse_comment_as_section(comment: &str) -> Option<(usize, String)> {
...
+            } else {
+                break;
+            }
+        }
 
-            // Return early to avoid further processing
-            return Ok(true);
+        // Push the new symbol to the current parent
+        if let Some((_, current_parent_ptr)) = section_stack.last() {
+            let current_parent = unsafe { &mut **current_parent_ptr };
```
Comment (kv9898) [inline crates/ark/src/lsp/symbols.rs:174]: 1st unsafe reference

File: crates/ark/src/lsp/symbols.rs [hunk 5]
```diff
@@ -188,11 +216,23 @@ fn index_node(
...
-                error!("{:?}", error);
+            if child.node_type() == NodeType::Comment {
+                let result: Result<bool> =
+                    index_assignment_with_comments(&child, contents, parent, &mut section_stack);
+                if let Err(error) = result {
+                    error!("{:?}", error);
+                }
+            } else {
+                if let Some((_, current_parent_ptr)) = section_stack.last() {
+                    let current_parent = unsafe { &mut **current_parent_ptr };
```
Comment (kv9898) [inline crates/ark/src/lsp/symbols.rs:230]: 2nd unsafe reference

**Comment by kv9898** [comment]
Closed as refactored to https://github.com/posit-dev/ark/pull/611.

**Pull request closed without merging**